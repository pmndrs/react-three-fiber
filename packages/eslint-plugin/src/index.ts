// THIS FILE WAS GENERATED DO NOT MODIFY BY HAND
// @command yarn codegen:eslint

import all from './configs/all'
import recommended from './configs/recommended'

export { default as rules } from './rules/index'

export const configs = {
  all,
  recommended,
}
