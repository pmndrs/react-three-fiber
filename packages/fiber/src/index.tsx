export * from './three-types'
import * as ReactThreeFiber from './three-types'
export { ReactThreeFiber }
export type {
  Intersection,
  Subscription,
  Dpr,
  Size,
  Viewport,
  Camera,
  RenderCallback,
  Performance,
  RootState,
} from './core/store'
export type { ThreeEvent, Events, EventManager } from './core/events'
export type { ObjectMap } from './core/utils'
export * from './web/Canvas'
export * from './web'
