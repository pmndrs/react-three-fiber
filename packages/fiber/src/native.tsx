export * from './three-types'
import * as ReactThreeFiber from './three-types'
export { ReactThreeFiber }
export type {
  Intersection,
  Subscription,
  Dpr,
  Size,
  Viewport,
  Camera,
  RenderCallback,
  Performance,
  RootState,
} from './core/store'
export type { ThreeEvent, Events, EventManager } from './core/events'
export type { ObjectMap } from './core/utils'
export * from './native/Canvas'
export * from './native/index'
