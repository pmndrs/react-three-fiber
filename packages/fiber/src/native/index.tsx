import * as THREE from 'three'
import * as React from 'react'
// @ts-ignore
import { ConcurrentRoot } from 'react-reconciler/constants'
import { UseStore } from 'zustand'

import { dispose } from '../core/utils'
import { Renderer, isRenderer, createStore, StoreProps, context, RootState, Size } from '../core/store'
import { createRenderer, extend, Root } from '../core/renderer'
import { createLoop, addEffect, addAfterEffect, addTail } from '../core/loop'
import { createTouchEvents as events, getEventPriority } from './events'
import { Canvas } from './Canvas'
import { EventManager } from '../core/events'
import { View, PixelRatio } from 'react-native'
import { ExpoWebGLRenderingContext } from 'expo-gl'

export type GLContext = ExpoWebGLRenderingContext | WebGLRenderingContext

type Properties<T> = Pick<T, { [K in keyof T]: T[K] extends (_: any) => any ? never : K }[keyof T]>

type GLProps =
  | Renderer
  | ((canvas: HTMLCanvasElement, context: GLContext) => Renderer)
  | Partial<Properties<THREE.WebGLRenderer> | THREE.WebGLRendererParameters>
  | undefined

const roots = new Map<GLContext, Root>()
const { invalidate, advance } = createLoop(roots)
const { reconciler, applyProps } = createRenderer(roots, getEventPriority)

export type RenderProps<TView extends View> = Omit<StoreProps, 'gl' | 'events' | 'size' | 'dpr'> & {
  gl?: GLProps
  events?: (store: UseStore<RootState>) => EventManager<TView>
  size?: Size
  onCreated?: (state: RootState) => void
}

const createRendererInstance = (gl: GLProps, context: GLContext): THREE.WebGLRenderer => {
  // Create canvas shim
  const canvas = {
    width: context.drawingBufferWidth,
    height: context.drawingBufferHeight,
    style: {},
    addEventListener: (() => {}) as any,
    removeEventListener: (() => {}) as any,
    clientHeight: context.drawingBufferHeight,
  } as HTMLCanvasElement

  // If a renderer is specified, return it
  const customRenderer = (typeof gl === 'function' ? gl(canvas, context) : gl) as THREE.WebGLRenderer
  if (isRenderer(customRenderer)) return customRenderer

  // Create renderer and pass our canvas and its context
  const renderer = new THREE.WebGLRenderer({
    powerPreference: 'high-performance',
    antialias: true,
    alpha: true,
    ...(gl as any),
    canvas,
    context,
  })

  // Set color management
  renderer.outputEncoding = THREE.sRGBEncoding
  renderer.toneMapping = THREE.ACESFilmicToneMapping

  // Set GL props
  if (gl) applyProps(renderer as any, gl as any)

  // Bind render to RN bridge
  if ((context as ExpoWebGLRenderingContext).endFrameEXP) {
    const renderFrame = renderer.render.bind(renderer)
    renderer.render = (scene: THREE.Scene, camera: THREE.Camera) => {
      renderFrame(scene, camera)
      ;(context as ExpoWebGLRenderingContext).endFrameEXP()
    }
  }

  return renderer
}

function createRoot<TView extends View>(context: GLContext, config?: RenderProps<TView>) {
  return {
    render: (element: React.ReactNode) => render(element, context, config),
    unmount: () => unmountComponentAtNode(context),
  }
}

function render<TView extends View>(
  element: React.ReactNode,
  context: GLContext,
  { gl, size = { width: 0, height: 0 }, events, onCreated, ...props }: RenderProps<TView> = {},
): UseStore<RootState> {
  let root = roots.get(context)
  let fiber = root?.fiber
  let store = root?.store
  let state = store?.getState()

  if (fiber && state) {
    // When a root was found, see if any fundamental props must be changed or exchanged

    // Check size
    if (state.size.width !== size.width || state.size.height !== size.height) state.setSize(size.width, size.height)

    // For some props we want to reset the entire root

    // Changes to the color-space
    const linearChanged = props.linear !== state.internal.lastProps.linear
    if (linearChanged) {
      unmountComponentAtNode(context)
      fiber = undefined
    }
  }

  if (!fiber) {
    // If no root has been found, make one

    // Create gl
    const glRenderer = createRendererInstance(gl, context)

    // Create store
    store = createStore(applyProps, invalidate, advance, {
      gl: glRenderer,
      size,
      ...props,
      // expo-gl can only render at native dpr/resolution
      // https://github.com/expo/expo-three/issues/39
      dpr: PixelRatio.get(),
    })
    const state = store.getState()
    // Create renderer
    fiber = reconciler.createContainer(store, ConcurrentRoot, false, null)
    // Map it
    roots.set(context, { fiber, store })
    // Store event manager internally and connect it
    if (events) {
      state.set({ events: events(store) })
      state.get().events.connect?.(context)
    }
  }

  if (store && fiber) {
    reconciler.updateContainer(
      <Provider store={store} element={element} onCreated={onCreated} />,
      fiber,
      null,
      () => undefined,
    )
    return store
  } else {
    throw 'Error creating root!'
  }
}

function Provider({
  store,
  element,
  onCreated,
}: {
  onCreated?: (state: RootState) => void
  store: UseStore<RootState>
  element: React.ReactNode
}) {
  React.useEffect(() => {
    const state = store.getState()
    // Flag the canvas active, rendering will now begin
    state.set((state) => ({ internal: { ...state.internal, active: true } }))
    // Notifiy that init is completed, the scene graph exists, but nothing has yet rendered
    if (onCreated) onCreated(state)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
  return <context.Provider value={store}>{element}</context.Provider>
}

function unmountComponentAtNode(context: GLContext, callback?: (context: GLContext) => void) {
  const root = roots.get(context)
  const fiber = root?.fiber
  if (fiber) {
    const state = root?.store.getState()
    if (state) state.internal.active = false
    reconciler.updateContainer(null, fiber, null, () => {
      if (state) {
        setTimeout(() => {
          try {
            state.events.disconnect?.()
            state.gl?.renderLists?.dispose?.()
            state.gl?.forceContextLoss?.()
            dispose(state)
            roots.delete(context)
            if (callback) callback(context)
          } catch (e) {
            /* ... */
          }
        }, 500)
      }
    })
  }
}

const act = reconciler.act
function createPortal(children: React.ReactNode, container: THREE.Object3D): React.ReactNode {
  return reconciler.createPortal(children, container, null, null)
}

reconciler.injectIntoDevTools({
  bundleType: process.env.NODE_ENV === 'production' ? 0 : 1,
  rendererPackageName: '@react-three/fiber',
  version: '18.0.0',
})

export * from './hooks'
export {
  context,
  render,
  createRoot,
  unmountComponentAtNode,
  createPortal,
  events,
  reconciler,
  applyProps,
  dispose,
  invalidate,
  advance,
  extend,
  addEffect,
  addAfterEffect,
  addTail,
  Canvas,
  act,
  roots as _roots,
}
