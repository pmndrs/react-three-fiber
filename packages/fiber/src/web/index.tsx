import * as THREE from 'three'
import * as React from 'react'
import { RootTag } from 'react-reconciler'
import { UseStore } from 'zustand'

import { is } from '../core/is'
import { createStore, StoreProps, isRenderer, context, RootState, Size, calculateDpr } from '../core/store'
import { createRenderer, extend, Root } from '../core/renderer'
import { createLoop, addEffect, addAfterEffect, addTail } from '../core/loop'
import { createPointerEvents as events } from './events'
import { Canvas } from './Canvas'
import { EventManager } from '../core/events'

const roots = new Map<Element, Root>()
const modes = ['legacy', 'blocking', 'concurrent'] as const
const { invalidate, advance } = createLoop(roots)
const { reconciler, applyProps } = createRenderer(roots)

export type RenderProps<TCanvas extends Element> = Omit<StoreProps, 'gl' | 'events' | 'size'> & {
  gl?: THREE.WebGLRenderer | Partial<THREE.WebGLRendererParameters>
  events?: (store: UseStore<RootState>) => EventManager<TCanvas>
  size?: Size
  mode?: typeof modes[number]
  onCreated?: (state: RootState) => void
}

const createRendererInstance = <TElement extends Element>(
  gl: THREE.WebGLRenderer | Partial<THREE.WebGLRendererParameters> | undefined,
  canvas: TElement,
): THREE.WebGLRenderer =>
  isRenderer(gl as THREE.WebGLRenderer)
    ? (gl as THREE.WebGLRenderer)
    : new THREE.WebGLRenderer({
        powerPreference: 'high-performance',
        canvas: canvas as unknown as HTMLCanvasElement,
        antialias: true,
        alpha: true,
        ...gl,
      })

function render<TCanvas extends Element>(
  element: React.ReactNode,
  canvas: TCanvas,
  { gl, size, mode = modes[1], events, onCreated, ...props }: RenderProps<TCanvas> = {},
): UseStore<RootState> {
  // Allow size to take on container bounds initially
  if (!size) {
    size = {
      width: canvas.parentElement?.clientWidth ?? 0,
      height: canvas.parentElement?.clientHeight ?? 0,
    }
  }

  let root = roots.get(canvas)
  let fiber = root?.fiber
  let store = root?.store
  let state = store?.getState()

  if (fiber && state) {
    // When a root was found, see if any fundamental props must be changed or exchanged

    // Check pixelratio
    if (props.dpr !== undefined && !is.equ(state.viewport.dpr, calculateDpr(props.dpr))) state.setDpr(props.dpr)
    // Check size
    if (state.size.width !== size.width || state.size.height !== size.height) state.setSize(size.width, size.height)

    // For some props we want to reset the entire root

    // Changes to the color-space
    const linearChanged = props.linear !== state.internal.lastProps.linear
    if (linearChanged) {
      unmountComponentAtNode(canvas)
      fiber = undefined
    }
  }

  if (!fiber) {
    // If no root has been found, make one

    // Create gl
    const glRenderer = createRendererInstance(gl, canvas)

    // Enable VR if requested
    if (props.vr) {
      glRenderer.xr.enabled = true
      glRenderer.setAnimationLoop((timestamp) => advance(timestamp, true))
    }

    // Create store
    store = createStore(applyProps, invalidate, advance, { gl: glRenderer, size, ...props })
    const state = store.getState()
    // Create renderer
    fiber = reconciler.createContainer(store, modes.indexOf(mode) as RootTag, false, null)
    // Map it
    roots.set(canvas, { fiber, store })
    // Store events internally
    if (events) state.set({ events: events(store) })
  }

  if (store && fiber) {
    reconciler.updateContainer(
      <Provider store={store} element={element} onCreated={onCreated} target={canvas} />,
      fiber,
      null,
      () => undefined,
    )
    return store
  } else {
    throw 'Error creating root!'
  }
}

function Provider<TElement extends Element>({
  store,
  element,
  onCreated,
  target,
}: {
  onCreated?: (state: RootState) => void
  store: UseStore<RootState>
  element: React.ReactNode
  target: TElement
}) {
  React.useEffect(() => {
    const state = store.getState()
    // Flag the canvas active, rendering will now begin
    state.set((state) => ({ internal: { ...state.internal, active: true } }))
    // Connect events
    state.events.connect?.(target)
    // Notifiy that init is completed, the scene graph exists, but nothing has yet rendered
    if (onCreated) onCreated(state)
  }, [])
  return <context.Provider value={store}>{element}</context.Provider>
}

function unmountComponentAtNode<TElement extends Element>(canvas: TElement, callback?: (canvas: TElement) => void) {
  const root = roots.get(canvas)
  const fiber = root?.fiber
  if (fiber) {
    const state = root?.store.getState()
    if (state) state.internal.active = false
    reconciler.updateContainer(null, fiber, null, () => {
      if (state) {
        setTimeout(() => {
          state.events.disconnect?.()
          state.gl?.renderLists?.dispose?.()
          state.gl?.forceContextLoss?.()
          dispose(state)
          roots.delete(canvas)
          if (callback) callback(canvas)
        }, 500)
      }
    })
  }
}

function dispose<TObj extends { dispose?: () => void; type?: string; [key: string]: any }>(obj: TObj) {
  if (obj.dispose && obj.type !== 'Scene') obj.dispose()
  for (const p in obj) {
    ;(p as any).dispose?.()
    delete obj[p]
  }
}

const act = reconciler.act
function createPortal(children: React.ReactNode, container: THREE.Object3D): React.ReactNode {
  return reconciler.createPortal(children, container, null, null)
}

reconciler.injectIntoDevTools({
  bundleType: process.env.NODE_ENV === 'production' ? 0 : 1,
  rendererPackageName: '@react-three/fiber',
  version: '17.0.2',
})

export * from '../core/hooks'
export {
  context,
  render,
  unmountComponentAtNode,
  createPortal,
  events,
  reconciler,
  applyProps,
  dispose,
  invalidate,
  advance,
  extend,
  addEffect,
  addAfterEffect,
  addTail,
  Canvas,
  act,
  roots as _roots,
}
