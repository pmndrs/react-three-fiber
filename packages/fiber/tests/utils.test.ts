import * as THREE from 'three'
import { Instance, RootStore } from '../src'
import {
  is,
  dispose,
  REACT_INTERNAL_PROPS,
  getInstanceProps,
  prepare,
  resolve,
  attach,
  detach,
  RESERVED_PROPS,
  diffProps,
  applyProps,
  updateCamera,
} from '../src/core/utils'

// Mocks a Zustand store
const storeMock: RootStore = Object.assign(() => null!, {
  getState: () => null!,
  setState() {},
  subscribe: () => () => {},
  destroy() {},
})

describe('is', () => {
  const myFunc = () => null
  const myObj = { myProp: 'test-prop' }
  const myStr = 'test-string'
  const myNum = 1
  const myUnd = undefined
  const myArr = [1, 2, 3]

  it('should tell me if something is a function', () => {
    expect(is.fun(myFunc)).toBe(true)

    expect(is.fun(myObj)).toBe(false)
    expect(is.fun(myStr)).toBe(false)
    expect(is.fun(myNum)).toBe(false)
    expect(is.fun(myUnd)).toBe(false)
    expect(is.fun(myArr)).toBe(false)
  })
  it('should tell me if something is an object', () => {
    expect(is.obj(myFunc)).toBe(false)

    expect(is.obj(myObj)).toBe(true)

    expect(is.obj(myStr)).toBe(false)
    expect(is.obj(myNum)).toBe(false)
    expect(is.obj(myUnd)).toBe(false)
    expect(is.obj(myArr)).toBe(false)
  })
  it('should tell me if something is a string', () => {
    expect(is.str(myFunc)).toBe(false)
    expect(is.str(myObj)).toBe(false)

    expect(is.str(myStr)).toBe(true)

    expect(is.str(myNum)).toBe(false)
    expect(is.str(myUnd)).toBe(false)
    expect(is.str(myArr)).toBe(false)
  })
  it('should tell me if something is a number', () => {
    expect(is.num(myFunc)).toBe(false)
    expect(is.num(myObj)).toBe(false)
    expect(is.num(myStr)).toBe(false)

    expect(is.num(myNum)).toBe(true)

    expect(is.num(myUnd)).toBe(false)
    expect(is.num(myArr)).toBe(false)
  })
  it('should tell me if something is undefined', () => {
    expect(is.und(myFunc)).toBe(false)
    expect(is.und(myObj)).toBe(false)
    expect(is.und(myStr)).toBe(false)
    expect(is.und(myNum)).toBe(false)

    expect(is.und(myUnd)).toBe(true)

    expect(is.und(myArr)).toBe(false)
  })
  it('should tell me if something is an array', () => {
    expect(is.arr(myFunc)).toBe(false)
    expect(is.arr(myObj)).toBe(false)
    expect(is.arr(myStr)).toBe(false)
    expect(is.arr(myNum)).toBe(false)
    expect(is.arr(myUnd)).toBe(false)

    expect(is.arr(myArr)).toBe(true)
  })
  it('should tell me if something is equal', () => {
    expect(is.equ([], '')).toBe(false)

    expect(is.equ('hello', 'hello')).toBe(true)
    expect(is.equ(1, 1)).toBe(true)

    const obj = { type: 'Mesh' }
    expect(is.equ(obj, obj)).toBe(true)
    expect(is.equ({}, {})).toBe(false)
    expect(is.equ({}, {}, { objects: 'reference' })).toBe(false)
    expect(is.equ({}, {}, { objects: 'shallow' })).toBe(true)
    expect(is.equ({ a: 1 }, { a: 1 })).toBe(false)
    expect(is.equ({ a: 1 }, { a: 1 }, { objects: 'reference' })).toBe(false)
    expect(is.equ({ a: 1 }, { a: 1 }, { objects: 'shallow' })).toBe(true)
    expect(is.equ({ a: 1, b: 1 }, { a: 1 }, { objects: 'shallow' })).toBe(false)
    expect(is.equ({ a: 1 }, { a: 1, b: 1 }, { objects: 'shallow' })).toBe(false)
    expect(is.equ({ a: 1 }, { a: 1, b: 1 }, { objects: 'shallow', strict: false })).toBe(true)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3] }, { arrays: 'reference', objects: 'reference' })).toBe(false)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3] }, { objects: 'reference' })).toBe(false)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3] }, { objects: 'shallow' })).toBe(true)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3, 4] }, { objects: 'shallow' })).toBe(false)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3, 4] }, { objects: 'shallow', strict: false })).toBe(true)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3], b: 1 }, { objects: 'shallow' })).toBe(false)
    expect(is.equ({ a: [1, 2, 3] }, { a: [1, 2, 3], b: 1 }, { objects: 'shallow', strict: false })).toBe(true)

    const arr = [1, 2, 3]
    expect(is.equ(arr, arr)).toBe(true)
    expect(is.equ([], [])).toBe(true)
    expect(is.equ([], [], { arrays: 'reference' })).toBe(false)
    expect(is.equ([], [], { arrays: 'shallow' })).toBe(true)
    expect(is.equ([1, 2, 3], [1, 2, 3])).toBe(true)
    expect(is.equ([1, 2, 3], [1, 2, 3], { arrays: 'shallow' })).toBe(true)
    expect(is.equ([1, 2, 3], [1, 2, 3], { arrays: 'reference' })).toBe(false)
    expect(is.equ([1, 2], [1, 2, 3])).toBe(false)
    expect(is.equ([1, 2, 3, 4], [1, 2, 3])).toBe(false)
    expect(is.equ([1, 2], [1, 2, 3], { strict: false })).toBe(true)
  })
})

describe('dispose', () => {
  it('should dispose of objects and their properties', () => {
    const mesh = Object.assign(new THREE.Mesh<THREE.BoxGeometry, THREE.MeshBasicMaterial>(), { dispose: jest.fn() })
    mesh.material.dispose = jest.fn()
    mesh.geometry.dispose = jest.fn()

    dispose(mesh)
    expect(mesh.dispose).toBeCalled()
    expect(mesh.material.dispose).toBeCalled()
    expect(mesh.geometry.dispose).toBeCalled()
  })

  it('should not dispose of a THREE.Scene', () => {
    const scene = Object.assign(new THREE.Scene(), { dispose: jest.fn() })

    dispose(scene)
    expect(scene.dispose).not.toBeCalled()

    const disposable = { dispose: jest.fn(), scene }
    dispose(disposable)
    expect(disposable.dispose).toBeCalled()
    expect(disposable.scene.dispose).not.toBeCalled()
  })
})

describe('getInstanceProps', () => {
  it('should filter internal props without accessing them', () => {
    const get = jest.fn()
    const set = jest.fn()

    const props = { foo: true }
    const filtered = getInstanceProps(
      REACT_INTERNAL_PROPS.reduce((acc, key) => ({ ...acc, [key]: { get, set } }), props),
    )

    expect(filtered).toStrictEqual(props)
    expect(get).not.toBeCalled()
    expect(set).not.toBeCalled()
  })
})

describe('prepare', () => {
  it('should create an instance descriptor', () => {
    const object = new THREE.Object3D()
    const instance = prepare(object, storeMock, 'object3D', { name: 'object' })

    expect(instance.root).toBe(storeMock)
    expect(instance.type).toBe('object3D')
    expect(instance.props.name).toBe('object')
    expect(instance.object).toBe(object)
    expect((object as Instance<THREE.Object3D>['object']).__r3f).toBe(instance)
  })

  it('should not overwrite descriptors', () => {
    const containerDesc = {}
    const container = { __r3f: containerDesc }

    const instance = prepare(container, storeMock, 'container', {})
    expect(container.__r3f).toBe(containerDesc)
    expect(instance).toBe(containerDesc)
  })
})

describe('resolve', () => {
  it('should resolve pierced props', () => {
    const object = { foo: { bar: 1 } }
    const { root, key, target } = resolve(object, 'foo-bar')

    expect(root).toBe(object['foo'])
    expect(key).toBe('bar')
    expect(target).toBe(root[key])
  })

  it('should switch roots for atomic targets', () => {
    const bar = new THREE.Vector3()
    const object = { foo: { bar } }
    const { root, key, target } = resolve(object, 'foo-bar')

    expect(root).toBe(object)
    expect(key).toBe('bar')
    expect(target).toBe(bar)
  })
})

describe('attach / detach', () => {
  it('should attach & detach using string values', () => {
    const parent = prepare({ prop: null }, storeMock, '', {})
    const child = prepare({}, storeMock, '', { attach: 'prop' })

    attach(parent, child)
    expect(parent.object.prop).toBe(child.object)
    expect(child.previousAttach).toBe(null)

    detach(parent, child)
    expect(parent.object.prop).toBe(null)
    expect(child.previousAttach).toBe(undefined)
  })

  it('should attach & detach using attachFns', () => {
    const mount = jest.fn()
    const unmount = jest.fn()

    const parent = prepare({}, storeMock, '', {})
    const child = prepare({}, storeMock, '', { attach: () => (mount(), unmount) })

    attach(parent, child)
    expect(mount).toBeCalledTimes(1)
    expect(unmount).toBeCalledTimes(0)
    expect(child.previousAttach).toBe(unmount)

    detach(parent, child)
    expect(mount).toBeCalledTimes(1)
    expect(unmount).toBeCalledTimes(1)
    expect(child.previousAttach).toBe(undefined)
  })

  it('should create array when using array-index syntax', () => {
    const parent = prepare({ prop: null }, storeMock, '', {})
    const child = prepare({}, storeMock, '', { attach: 'prop-0' })

    attach(parent, child)
    expect(parent.object.prop).toStrictEqual([child.object])
    expect(child.previousAttach).toBe(undefined)

    detach(parent, child)
    expect((parent.object.prop as unknown as Array<never>).length).toBe(1)
    expect((parent.object.prop as unknown as Array<never>)[0]).toBe(undefined)
    expect(child.previousAttach).toBe(undefined)
  })
})

describe('diffProps', () => {
  it('should filter changed props', () => {
    const instance = prepare({}, storeMock, '', { foo: true })
    const newProps = { foo: true, bar: false }

    const filtered = diffProps(instance, newProps)
    expect(filtered).toStrictEqual({ bar: false })
  })

  it('invalidates pierced props when root is changed', async () => {
    const texture1 = { needsUpdate: false, name: '' } as THREE.Texture
    const texture2 = { needsUpdate: false, name: '' } as THREE.Texture

    const oldProps = { map: texture1, 'map-needsUpdate': true, 'map-name': 'test' }
    const newProps = { map: texture2, 'map-needsUpdate': true, 'map-name': 'test' }

    const instance = prepare({}, storeMock, '', oldProps)
    const filtered = diffProps(instance, newProps)
    expect(filtered).toStrictEqual(newProps)
  })

  it('should pick removed props for HMR', () => {
    const instance = prepare(new THREE.Object3D(), storeMock, '', { position: [0, 0, 1] })
    const newProps = {}

    const filtered = diffProps(instance, newProps, true)
    expect(filtered).toStrictEqual({ position: new THREE.Object3D().position })
  })

  it('should reset removed props for HMR', () => {
    const instance = prepare(new THREE.Object3D(), storeMock, '', { scale: 10 })
    const filtered = diffProps(instance, {}, true)
    expect((filtered.scale as THREE.Vector3).toArray()).toStrictEqual([1, 1, 1])
  })

  it('should filter reserved props without accessing them', () => {
    const get = jest.fn()
    const set = jest.fn()

    const props = { foo: true }
    const filtered = diffProps(
      prepare({}, storeMock, '', {}),
      RESERVED_PROPS.reduce((acc, key) => ({ ...acc, [key]: { get, set } }), props),
    )

    expect(filtered).toStrictEqual(props)
    expect(get).not.toBeCalled()
    expect(set).not.toBeCalled()
  })
})

describe('applyProps', () => {
  it('should apply props to foreign objects', () => {
    const target = new THREE.Object3D()
    expect(() => applyProps(target, {})).not.toThrow()
  })

  it('should filter reserved props without accessing them', () => {
    const get = jest.fn()
    const set = jest.fn()

    const props = { foo: true }
    const target = {}
    applyProps(
      target,
      RESERVED_PROPS.reduce((acc, key) => ({ ...acc, [key]: { get, set } }), props),
    )

    expect(target).toStrictEqual(props)
    expect(get).not.toBeCalled()
    expect(set).not.toBeCalled()
  })

  it('should overwrite non-atomic properties', () => {
    const foo = { value: true }
    const target = { foo }
    applyProps(target, { foo: { value: false } })

    expect(target.foo).not.toBe(foo)
    expect(target.foo.value).toBe(false)
  })

  it('should prefer to copy from external props', async () => {
    const color = new THREE.Color()
    color.copy = jest.fn()

    const target = { color, layer: new THREE.Layers() }

    // Same constructor, copy
    applyProps(target, { color: new THREE.Color() })
    expect(target.color).toBeInstanceOf(THREE.Color)
    expect(color.copy).toHaveBeenCalledTimes(1)

    // Same constructor, Layers
    const layer = new THREE.Layers()
    layer.mask = 5
    applyProps(target, { layer })
    expect(target.layer).toBeInstanceOf(THREE.Layers)
    expect(target.layer.mask).toBe(layer.mask)

    // Different constructor, overwrite
    applyProps(target, { color: new THREE.Vector3() })
    expect(target.color).toBeInstanceOf(THREE.Vector3)
    expect(color.copy).toHaveBeenCalledTimes(1)
  })

  it('should prefer to set when props are an array', async () => {
    const target = new THREE.Object3D()
    applyProps(target, { position: [1, 2, 3] })

    expect(target.position.toArray()).toStrictEqual([1, 2, 3])
  })

  it('should set with scalar shorthand where applicable', async () => {
    // Vector3#setScalar
    const target = new THREE.Object3D()
    applyProps(target, { scale: 5 })
    expect(target.scale.toArray()).toStrictEqual([5, 5, 5])

    // Color#set
    const material = new THREE.MeshBasicMaterial()
    applyProps(material, { color: 0x000000 })
    expect(material.color.getHex()).toBe(0x000000)

    // No-op on undefined
    const mesh = new THREE.Mesh()
    applyProps(mesh, { position: undefined })
    expect(mesh.position.toArray()).toStrictEqual([0, 0, 0])
  })

  it('should pierce into nested properties', () => {
    const target = new THREE.Mesh<THREE.BufferGeometry, THREE.MeshBasicMaterial>()
    applyProps(target, { 'material-color': 0x000000 })

    expect(target.material.color.getHex()).toBe(0x000000)
  })

  it('should not apply a prop if it is undefined', async () => {
    const target = { value: 'initial' }
    applyProps(target, { value: undefined })

    expect(target.value).toBe('initial')
  })
})

describe('updateCamera', () => {
  it('updates camera matrices', () => {
    const size = { width: 1280, height: 800, left: 0, top: 0 }

    const perspective = new THREE.PerspectiveCamera()
    perspective.updateProjectionMatrix = jest.fn()
    perspective.updateMatrixWorld = jest.fn()
    updateCamera(perspective, size)
    expect(perspective.updateProjectionMatrix).toBeCalled()
    expect(perspective.updateMatrixWorld).toBeCalled()
    expect(perspective.projectionMatrix.toArray()).toMatchSnapshot()
    expect(perspective.matrixWorld.toArray()).toMatchSnapshot()

    const orthographic = new THREE.OrthographicCamera()
    orthographic.updateProjectionMatrix = jest.fn()
    orthographic.updateMatrixWorld = jest.fn()
    updateCamera(orthographic, size)
    expect(orthographic.updateProjectionMatrix).toBeCalled()
    expect(orthographic.updateMatrixWorld).toBeCalled()
    expect(orthographic.projectionMatrix.toArray()).toMatchSnapshot()
    expect(orthographic.matrixWorld.toArray()).toMatchSnapshot()
  })

  it('respects camera.manual', () => {
    const size = { width: 0, height: 0, left: 0, top: 0 }

    const perspective = Object.assign(new THREE.PerspectiveCamera(), { manual: true })
    perspective.updateProjectionMatrix = jest.fn()
    perspective.updateMatrixWorld = jest.fn()
    updateCamera(perspective, size)
    expect(perspective.updateProjectionMatrix).not.toBeCalled()
    expect(perspective.updateMatrixWorld).not.toBeCalled()

    const orthographic = Object.assign(new THREE.OrthographicCamera(), { manual: true })
    orthographic.updateProjectionMatrix = jest.fn()
    orthographic.updateMatrixWorld = jest.fn()
    updateCamera(orthographic, size)
    expect(orthographic.updateProjectionMatrix).not.toBeCalled()
    expect(orthographic.updateMatrixWorld).not.toBeCalled()
  })
})
