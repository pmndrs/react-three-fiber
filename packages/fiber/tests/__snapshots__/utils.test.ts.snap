// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateCamera updates camera matrices 1`] = `
Array [
  2.1445069205095586,
  0,
  0,
  0,
  0,
  2.1445069205095586,
  0,
  0,
  0,
  0,
  -1.00010000500025,
  -1,
  0,
  0,
  -0.200010000500025,
  0,
]
`;

exports[`updateCamera updates camera matrices 2`] = `
Array [
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`updateCamera updates camera matrices 3`] = `
Array [
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  -0.001000050002500125,
  0,
  -0,
  -0,
  -1.00010000500025,
  1,
]
`;

exports[`updateCamera updates camera matrices 4`] = `
Array [
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;
